sycl-fusion/passes/kernel-fusion/SYCLSpecConstMaterializer.cpp

```diff
@@ -0,0 +1,317 @@
+  auto *StringMD = MDString::get(
+      Context, std::string{(const char *)SpecConstData, SpecConstDataSize});
+  auto *TupleMD = MDTuple::get(Context, {StringMD});
+  SYCLMD->addOperand(TupleMD);
+
+  return PreservedAnalyses::all();
+}
+
+static Constant *getScalarConstant(const unsigned char **ValPtr, Type *Ty) {
+  if (Ty->isIntegerTy()) {
+    unsigned NumBytes = Ty->getIntegerBitWidth() / 8;
+    uint64_t IntValue = 0;
+    std::memcpy(&IntValue, *ValPtr, NumBytes);
+    *ValPtr = *ValPtr + NumBytes;
+    return ConstantInt::get(Ty, IntValue);
+  }
+  if (Ty->isDoubleTy()) {
+    double DoubleValue = *(reinterpret_cast<const double *>(*ValPtr));
+    *ValPtr = *ValPtr + sizeof(double);
+    return ConstantFP::get(Ty, DoubleValue);
+  }
+  if (Ty->isFloatTy()) {
+    float FloatValue = *(reinterpret_cast<const float *>(*ValPtr));
+    *ValPtr = *ValPtr + sizeof(float);
+    return ConstantFP::get(Ty, FloatValue);
+  }
+  if (Ty->isHalfTy()) {
+    uint16_t HalfValue = *(reinterpret_cast<const uint16_t *>(*ValPtr));
+    *ValPtr = *ValPtr + sizeof(uint16_t);
+    return ConstantFP::get(Ty, HalfValue);
+  }
+
+  llvm_unreachable("Scalar type not found.");
+}
+
+static Constant *getConstantOfType(const unsigned char **ValPtr, Type *Ty) {
+  assert(ValPtr && Ty && "Invalid input.");
+  if (Ty->isIntegerTy() || Ty->isFloatTy() || Ty->isDoubleTy() ||
+      Ty->isHalfTy())
+    return getScalarConstant(ValPtr, Ty);
+  if (auto *ArrTy = dyn_cast<ArrayType>(Ty)) {
+    SmallVector<Constant *> Elems;
+    auto *ElemTy = ArrTy->getArrayElementType();
+    auto NumElems = ArrTy->getArrayNumElements();
+    for (uint64_t I = 0; I < NumElems; ++I)
+      Elems.push_back(getConstantOfType(ValPtr, ElemTy));
+    return ConstantArray::get(ArrayType::get(ElemTy, NumElems), Elems);
+  }
+  if (auto *StructTy = dyn_cast<StructType>(Ty)) {
+    SmallVector<Constant *> StructElems;
+    for (auto *ElemTy : StructTy->elements())
+      StructElems.push_back(getConstantOfType(ValPtr, ElemTy));
+    return ConstantStruct::get(StructTy, StructElems);
+  }
+
+  llvm_unreachable("Unknown type in getConstantOfType.");
+}
+
+void SYCLSpecConstMaterializer::allocateSpecConstant(StringRef KernelName) {
+  for (auto I : llvm::enumerate(TypesAndOffsets)) {
+    auto *const Ty = I.value().first;
+    assert(Ty->isSized());
+    const auto Offset = I.value().second;
+    assert(Offset < SpecConstDataSize && "Out of bounds access.");
+    const unsigned char *ValPtr = &SpecConstData[Offset];
+    auto *Initializer = getConstantOfType(&ValPtr, I.value().first);
+    // AMD's CONSTANT_ADDRESS and Nvidia's ADDRESS_SPACE_CONST happen to have
+    // the same value.
+    const unsigned AS = 4;
+    auto *SpecConstGlobal = new GlobalVariable(
+        *Mod, Ty, /*isConstant*/ true, GlobalValue::WeakODRLinkage, Initializer,
+        Twine("SpecConsBlob_" + std::string(KernelName) + "_" +
+              std::to_string(I.index())),
```

Comment:
`Twine` can be used to avoid this string concatenation https://bcain-llvm.readthedocs.io/projects/llvm/en/latest/ProgrammersManual/#the-twine-class